### DISCUSSION TITLE
How to properly integrate esbuild bundle

### DISCUSSION THREAD

**Discussion by msorens** [microsoft/vscode-discussions#449] [Extension Development QnA] [answered]
As my new extension has progressed it reached a size where `vsce package` suggested: "For performance reasons, you should bundle your extension: https://aka.ms/vscode-bundle-extension ."
Visiting the suggested documentation, however, left me hanging. I wanted to use `esbuild`. The doc page provides the necessary entries for package.json/scripts...
```
"scripts": {
    "vscode:prepublish": "npm run esbuild-base -- --minify",
    "esbuild-base": "esbuild ./src/extension.ts --bundle --outfile=out/main.js --external:vscode --format=cjs --platform=node",
    "esbuild": "npm run esbuild-base -- --sourcemap",
    "esbuild-watch": "npm run esbuild-base -- --sourcemap --watch",
    "test-compile": "tsc -p ./"
},
```

... but it says nothing about how to integrate esbuild into the full build process. There are two integration points I need:

(1) Creating a package with `vsce package`. 

It builds the bundle (out/main.js) but there is clearly a piece missing to make it use the bundle (instead of the default extension.js):
```
% vsce package --pre-release
Executing prepublish script 'npm run vscode:prepublish'...

> vscode-trial@0.0.0-next-esbuild vscode:prepublish
> npm run esbuild-base -- --minify

> vscode-trial@0.0.0-next-esbuild esbuild-base
> esbuild ./src/extension.ts --bundle --outfile=out/main.js --external:vscode --format=cjs --platform=node --minify

  out/main.js  364.8kb

⚡ Done in 23ms
 ERROR  Extension entrypoint(s) missing. Make sure these files exist and aren't ignored by '.vscodeignore':
  extension/out/extension.js
```

(2) Debugging in VSCode.
If I leave the tasks.json as I had it before introducing esbuild, then pressing F5 populates the `out` directory with all the individual files so the extension works fine in the new "extension development host" window.
But I would have thought that I should be using the same bundle here (main.js), so I tried (via https://code.visualstudio.com/docs/editor/tasks) creating a fresh tasks.json via `Tasks: Configure Default Build Task` and selected `esbuild-watch`. This had two problems, though:
(a) it did not automatically open a new "extension development host" window until I somehow stumbled upon pressing `^C` in the terminal window and then it did!
(b) Once I got the "extension development host" window, the command palette did show my extension's commands, but when I tried to invoke one, it had essentially the same complaint about missing code; it could not find extension.js. It should have been looking in main.js, though, is what I thought.

**Comment by r3m0t** [comment, marked as answer]
Yes, there's a webpack example but no esbuild example - https://github.com/Microsoft/vscode-extension-samples/tree/main/webpack-sample

(1) change the entrypoint in the package.json to match where esbuild is putting the file.

(2) (a) you need to configure a problem matcher to tell VS Code when your build task has finished successfully. VS Code then starts the launch task. Otherwise, the default in VS Code is that your build task only finishes when it terminates, which in the case of a watch task is when you ^C it.

**Reply by msorens** [reply]
Marking as the correct answer even though you left me some work to do 😉 .
(1) Spot on! I had missed that that was specified in the package.json.
(2) Yes and no. Having no idea about how to configure one, a web search revealed there is an [esbuild problem matcher](https://marketplace.visualstudio.com/items?itemName=connor4312.esbuild-problem-matchers) extension available! The docs on that page gave the base content for tasks.json but that did not quite match up with what the VSCode docs on bundling suggested for esbuild... so I figured out a bit more about vscode tasks and then, with some experimentation, was able to tweak the tasks.json to sync with package.json and get F5 working properly for me. For other folks' reference, here is the tasks.json I derived:
```
{
	"version": "2.0.0",
	"tasks": [
		{
			"type": "npm",
			"script": "esbuild-watch",
			"group": {
				"kind": "build",
				"isDefault": true
			},
			"problemMatcher": "$esbuild-watch",
			"isBackground": true,
			"label": "npm: esbuild-watch",
			"detail": "npm run esbuild-base -- --sourcemap --watch"
		},
		{
			"type": "npm",
			"script": "esbuild",
			"group": "build",
			"problemMatcher": "$esbuild",
			"label": "npm: esbuild"
		}
	]
}
```

Thanks for the jump start!

**Reply by jakereifer** [reply]
What about debugging the tests? do you have to both run `test-compile` and `esbuild` so that not only are the (integration) tests built but also the extension itself can be loaded into the vscode test instance @msorens?
